File: LeetCode/src/main/java/com/jack/tree/LeetCode144.py

from typing import List, Optional
from TreeNode import TreeNode


class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        
        if root is None:
            return
        
        self.preorder(root, res)
        return res
        
        
    def preorder(self, node: Optional[TreeNode], res: List[int]):
        if node is None: return 
        
        res.append(node.val)
        self.preorder(node.left, res)
        self.preorder(node.right, res)
        


if __name__ == "__main__":
    pass

File: LeetCode/src/main/java/com/jack/tree/LeetCode144.java
package com.jack.tree;

import java.util.ArrayList;
import java.util.List;

public class LeetCode144 {
    public static void main(String[] args) {
        
    }


    static class Solution{
        public List<Integer> preorderTraversal(TreeNode root) {
            List<Integer> res = new ArrayList<>();
            if(root == null) return res;

            preorder(res, root);
            return res;
        
        }

        private void preorder(List<Integer> res, TreeNode node){
            if (node == null) return;
            
            res.add(node.val);
            preorder(res, node.left);
            preorder(res, node.right);
        }
    }

}
